### PR TITLE
Add dockerfile and compilation in CI/CD

### PR DESCRIPTION
**Pull request by jyothishnt**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM rust:1.67 as builder
```
Comment (jonk-ebi) [inline Dockerfile]: Have you tried switching to 1.67.1-alpine?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+RUN apt update
```
Comment (jonk-ebi) [inline Dockerfile]: This isn't needed as you are using the cargo package manager

File: Dockerfile (hunk 1 above)
Comment (kamaldodiya) [reply Dockerfile]: Yes, we did try it but we were getting some build error https://gitlab.ebi.ac.uk/ensembl-web/peregrine-eard/-/jobs/1248849  so we reverted to using rust:1.67

Comment (jonk-ebi) [reply Dockerfile]: I was able to build locally for alpine but needed to add 
`RUN apk add --no-cache musl-dev`

File: Dockerfile (hunk 2 above)
Comment (jonk-ebi) [reply Dockerfile:3]: Might not be worth the effort as it only saves <1GB

eard-debian       latest    5d7446f83b66   21 seconds ago   3.84GB
eard-alpine        latest    caaf1a478f55      8 minutes ago     3.09GB

File: Dockerfile (hunk 1 above)
Comment (jonk-ebi) [reply Dockerfile]: Might not be worth the effort as it only saves <1GB

eard-debian latest 5d7446f83b66 21 seconds ago 3.84GB
eard-alpine latest caaf1a478f55 8 minutes ago 3.09GB

Comment (jonk-ebi) [reply Dockerfile]: By splitting the build and the final image you can get the size down to 863Mb 
eard-alpine          latest    1ef21b56f17c   26 seconds ago   863MB

```
FROM rust:1.67.1-alpine as builder

RUN apk add --no-cache musl-dev

RUN rm -rf /app
RUN mkdir /app
COPY . /app

WORKDIR /app/compiler

RUN cargo build --release

FROM rust:1.67.1-alpine as run

RUN mkdir /app

COPY --from=builder /app/compiler/target/release/eard-compiler /app

ENV PATH=$PATH:/app
```

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Define stages for the CI/CD pipeline
+stages:
+  - build_docker_image
+
+variables:
+  CONTAINER_IMAGE: ${CI_REGISTRY_IMAGE}:eard-${CI_COMMIT_SHORT_SHA}
```
Comment (jonk-ebi) [inline .gitlab-ci.yml:6]: This will build a new image for every commit regardless of branch. It might make sense to limit it to main?

Comment (jonk-ebi) [reply .gitlab-ci.yml:6]: Actually this will disrupt the backend repo if it is using :latest

File: Dockerfile (hunk 1 above)
Comment (jonk-ebi) [reply Dockerfile]: If you change the second image from rust:alpine to just alpine you can get the image size down to 14.2Mb 🎉 
`FROM alpine:3.16 as run`

eard-alpine          latest          584a12d975ef   11 seconds ago      14.2MB

Comment (jyothishnt) [reply Dockerfile]: Wow! ok I will update and check again.

**Review by jonk-ebi — APPROVED** [review]
Looks good, great work 🎉

**Comment by kamaldodiya** [comment]
We still need to do a final check before we merge this to main. Let's just make sure that latest build earn-complier with alpine is working.

**Comment by jonk-ebi** [comment]
Local test of the new docker image 

I tested the new docker locally by doing the following
```
docker pull dockerhub.ebi.ac.uk/ensembl-web/peregrine-eard:latest
rm egs-data/begs/render16.eardo 
bash docker-begs.sh                                       
/Users/jon/Programming/ensembl-dauphin-style-compiler/backend-server
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
ls egs-data/begs 
boot10.begs		boot9.begs		render14.begs		specs16			stick8.begs
boot11.begs		program-inventory.toml	render15.begs		stick10.begs		stick9.begs
boot12.begs		render10.begs		**render16.eardo**		stick11.begs		test15.begs
boot13.begs		render11.begs		render8.begs		stick12.begs		test16.begs
boot14.begs		render12.begs		render9.begs		stick13.begs
boot8.begs		render13.begs		specs15			stick14.begs
```

docker-begs.sh is
```
#! /bin/bash

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
echo $DIR
export PATH="$PATH:$DIR/../dauphin/target/release"
#export PATH="$PATH:$DIR/../../peregrine-eard/compiler/target/release"

SRC="$DIR/egs-data/egs"
DST="$DIR/egs-data/begs"

#dauphin -c $SRC/v16/test.egs -c $SRC/v16/test-with-data.egs -o $DST/test16.begs -L peregrine -O2 
docker run --rm -v $DIR:$DIR dockerhub.ebi.ac.uk/ensembl-web/peregrine-eard:latest eard-compiler \
    -c $SRC/v16/other/framing.eard -c $SRC/v16/other/gc.eard -c $SRC/v16/other/contig.eard \
    -c $SRC/v16/other/zoomed-seq.eard -c $SRC/v16/variant/variant-summary.eard \
    -c $SRC/v16/other/ruler.eard -c $SRC/v16/gene/gene-overview.eard -c $SRC/v16/gene/gene.eard \
    -c $SRC/v16/gene/focus-gene.eard -c $SRC/v16/gene/focus-gene-dots.eard \
    -c $SRC/v16/gene/transcript.eard  -c $SRC/v16/other/focus-region.eard \
    -c $SRC/v16/gene/focus-transcript.eard -c $SRC/v16/gene/zoomed-transcript.eard \
    -c $SRC/v16/gene/focus-zoomed-transcript.eard -c $SRC/v16/other/focus-region.eard \
    -c $SRC/v16/variant/focus-variant.eard -c $SRC/v16/variant/focus-variant-summary.eard \
    -c $SRC/v16/variant/focus-variant-dots.eard \
    -o $DST/render16.eardo
```

**Review by kamaldodiya — APPROVED** [review]

**Pull request merged**